### ISSUE TITLE
POC TypeSpec to AsyncAPI

### ISSUE THREAD

**Issue by mikekistler** [microsoft/typespec#2463]
AsyncAPI is a sister specification to OpenAPI for describing asynchronous (e.g. message queuing) APIs. AsyncAPI is gaining some traction in the industry and by producing an emitter for this we could demonstrate the flexibility and power of TypeSpec in another API domain.

**Comment by allenjzhang** [comment]
Backlog

**Comment by SnakeDoc** [comment]
This feature would make TypeSpec a "do-all" tool for those writing OpenAPI and AsyncAPI specs for their teams. It's not uncommon to have both within a single system.

Adding my support for this feature! 👍

**Comment by kennethaasan** [comment]
I agree with the above. At Sportradar we use both OpenAPI and AsyncAPI with JSON Schema and would consider TypeSpec if AsyncAPI was supported.

**Comment by sandrolain** [comment]
I'm considering using typespec, but at the moment I'm mostly working on event driven architectures. Support for the generation of specifications for asynchronous APIs and event schemas, such as AsyncAPI and cloudevents, would also be fantastic.

**Comment by bterlson** [comment]
Folks chiming in for AsyncAPI support, please feel free to link your AsyncAPI specs if they are public, it would be great to take a look!

**Comment by chiodonia** [comment]
@SwissPost we designed hundreds of APIs (both REST and Streaming) using https://github.com/swisspost/apikana. Found TypeSpec very interesting but still missing the support for AsyncAPI

**Comment by SnakeDoc** [comment]
> Folks chiming in for AsyncAPI support, please feel free to link your AsyncAPI specs if they are public, it would be great to take a look!

I don't have a public spec to offer at the moment - however the AsyncAPI team does maintain example specs in their repos for testing, such as [this](https://github.com/asyncapi/bundler/tree/master/example) - just note the v3 directory contains AsyncAPI 3.x spec examples.

You may also be interested in these two discussions: [#141](https://github.com/asyncapi/bundler/issues/141) and [#151], which showcase some 3.x spec examples and expected outputs from the validation/bundling process.

**Comment by creatorrr** [comment]
any updates on this @bterlson ?

**Comment by bterlson** [comment]
Nothing new to share, unfortunately. Keep chiming in with support, though!

**Comment by SnakeDoc** [comment]
I'm chiming in with support! ;)

**Comment by aeworxet** [comment]
There are official examples in https://github.com/asyncapi/spec/tree/master/examples, and you can view an example for nearly any version of the AsyncAPI Specification:

```bash
$ git tag

1.1.0
1.2.0
2.0.0
...
v2.0.0
v2.1.0
...
v2.4.0
v2.4.0-2022-04-release.1
...
v2.5.0-next-spec.5
v2.6.0
v3.0.0
v3.0.0-next-major-spec.1
...

```

There is also a separate repository, which provides all the JSON Schema documents for validating AsyncAPI documents.
https://github.com/asyncapi/spec-json-schemas

**Comment by Kreash** [comment]
I'm waiting for support, I really need it

**Comment by LievenMasschelein-code** [comment]
Hi , we are also using both openapi and asyncapi ! asyncapi support in typespec would really make it a valid case to use typespec ! 
 I'm chimming in with support @bterlson

**Comment by mario-guerra** [comment]
We would like to support AsyncAPI but it doesn't map cleanly to TypeSpec in its current form. It would help to know how you would use AsyncAPI in practice. @LievenMasschelein-code @Kreash @chiodonia @SnakeDoc @kennethaasan @sandrolain @aeworxet

**Comment by aeworxet** [comment]
@mario-guerra

You can join [AsyncAPI Slack Workspace](https://asyncapi.com/slack-invite) and ask your questions in the channel
```
# 03_specification
4,403 members
All around the spec discussions. It is ok to ask for support here.
```

**Comment by SnakeDoc** [comment]
@mario-guerra Pretty much exactly the same way we use OpenAPI - our AsyncAPI spec is our "source of truth" for describing how our services should interact within our EDA.

I would guess one of the reasons you are not receiving many public AsyncAPI specs to look at is EDA/Message Broker stuff is often used internally.

That said, @aeworxet is a significant contributor to the AsyncAPI project and is a tremendous asset to have available for questioning! 🚀

**Comment by LievenMasschelein-code** [comment]
Hi for us also , just alike @SnakeDoc mentions,  we use in the same way we use Openapi. For us Asyncapi  is the only source of truth for our internal domain events between our microservices, we are building some automation on top of Asyncapi to have data quality checks for events and messages ingested in our data lakehouse platform. Asyncapi is also used to add events schema's in out confluent cloud ( kafka) schema registry and to enforce them.

**Comment by aeworxet** [comment]
> @aeworxet is a significant contributor to the AsyncAPI project and is a tremendous asset

😊

@fmvilas is the Founder of AsyncAPI, and @derberg is the Executive Director of AsyncAPI, so they are much more tremendous assets for asking questions. 🙂

**Comment by chiodonia** [comment]
Hi, we at SwissPost have been developing https://github.com/swisspost/apikana which is similar to TypeSpec. 
Apikana has been used to design hundreds of REST and Stream APIs (Kafka and MQTT). 

We are now willing to adopt asyncAPI for our stream APIs and are considering generating asyncAPI with Apikana. Since Apikana needs some investments we are also considering TypeSpec and are very interested in asyncAPI PoC.

From a stream API design point of view, we need to define the following asyncAPI elements:
- Info with additional x- fields
- Channels (topic name) with additional x- fields
- Messages and payloads (Json schema)
- Operations

I guess bindings, and other serialization formats (Protobuf,...) could be added later after the PoC.

AsyncAPI Servers and other elements are probably not relevant during the API design phase. The generated asyncAPI descriptors could be enriched with those elements when topics are deployed.

Andrea

**Comment by TertiumOrganum1** [comment]
Yes, very interested in asyncapi support too! Especially ver 3.0,which is much clearer and richer than 2.x.

**Comment by SnakeDoc** [comment]
Can anyone at Microsoft give us a hint if this is being worked on? The demand for this feature is here! A little extra transparency about status and ETA would be much appreciated.

**Comment by vec715** [comment]
it would be really nice to have AsyncAPI support!

**Comment by SnakeDoc** [comment]
Well, we're having to "bite the bullet" and move on without TypeSpec, which means we're probably not going to circle back. I hate to be negative, but it seems the Microsoft team vastly overstated the flexibility and power of their emitter architecture... both OpenAPI and AsyncAPI are based on JSON Schema, and the demand for this feature is apparent. The lack of communication makes me believe this is a very challenging/impossible task as currently written. Serious bummer...

**Comment by fmvilas** [comment]
@mario-guerra just writing to let you know that I'm available to help moving this forward. From understanding the spec inside out to getting in touch with implementers so they can tell a better story. Hit me on LinkedIn: https://www.linkedin.com/in/fmvilas.

**Comment by LarsArtmann** [comment]
🎯 **AsyncAPI 3.x Emitter Implementation Plan**

Hi Microsoft TypeSpec team! 👋

I'm working on implementing the AsyncAPI 3.x emitter that the community has been requesting. Based on the 37 👍 reactions here, there's clearly strong demand for this feature.

## 🏗️ **Current Progress**

I've created a comprehensive implementation plan:
- **GitHub Issues**: https://github.com/LarsArtmann/typespec-eventsourcing/issues/38
- **Technical Breakdown**: https://github.com/LarsArtmann/typespec-eventsourcing/issues/39

## 📋 **Implementation Approach**

### AsyncAPI 3.x Document Structure
```yaml
asyncapi: 3.0.0
info:
  title: Generated from TypeSpec
  version: 1.0.0
channels:
  userEvents:
    address: ecommerce.usermanagement.events
    messages:
      UserCreated:
        $ref: '#/components/messages/UserCreated'
operations:
  sendUserCreated:
    action: send
    channel:
      $ref: '#/channels/userEvents'
```

### Convention-Based Detection
```typescript
// TypeSpec → AsyncAPI mapping
namespace ECommerce.UserManagement {
  model UserCreated {     // → AsyncAPI message + send operation
    userId: string;
    email: string;
  }
  
  model CreateUser {      // → AsyncAPI message + receive operation
    email: string;
    name: string;
  }
}
```

## 🎯 **Key Features**

1. **Zero Setup Required** - Convention-based detection like the community wants
2. **AsyncAPI 3.x Compliance** - Full support for latest specification
3. **TypeSpec Integration** - Leverages existing JSON Schema generation
4. **Enterprise Ready** - Comprehensive error handling and validation

## 🤝 **Collaboration Opportunity**

I'd love to collaborate with the Microsoft team on making this official! I'm planning to:

1. **Build reference implementation** with comprehensive test coverage
2. **Get community feedback** and iterate based on real-world usage
3. **Submit PR for official consideration** once it's proven and stable

## 📚 **References**

- **AsyncAPI 3.x Spec**: https://www.asyncapi.com/docs/reference/specification/v3.0.0
- **Implementation Repo**: https://github.com/LarsArtmann/typespec-eventsourcing
- **EventCatalog Integration**: Already working on TypeSpec → EventCatalog bridge

## ❓ **Questions for Microsoft Team**

1. Are there specific AsyncAPI patterns you'd like to see prioritized?
2. Any preferences for how TypeSpec namespaces should map to AsyncAPI channels?
3. Would you be open to reviewing a reference implementation?

**This could make TypeSpec the go-to solution for event-driven architecture documentation!** 🚀

CC: @bterlson (as mentioned in the issue) - would love your thoughts on the approach!

---

*Generated with ❤️ by a passionate TypeSpec community member*

**Comment by bZichett** [comment]
Bumping.  Would like to see this 🙏

**Comment by LarsArtmann** [comment]
>  [!NOTE]
> I am currently developing a TypeSpec AsyncAPI emitter prototype independently with the goal of potentially contributing it to this repository.

@fmvilas Quick question about AsyncAPI versioning capabilities:

Does AsyncAPI support native multi-version document generation like OpenAPI 3.x?

From my research, AsyncAPI has:
✅ info.version field
✅ deprecated: true flags
✅ Schema composition (allOf, oneOf)

But lacks:
❌ Multi-version document generation
❌ Built-in version evolution mechanisms
❌ Official version migration tooling

My plan: Generate separate AsyncAPI files per version (asyncapi.v1.yaml, asyncapi.v2.yaml) using
TypeSpec's `@added`/`@removed` decorators.

Is this the right approach, or does AsyncAPI have native versioning I missed?

**Comment by TertiumOrganum1** [comment]
Hi!
Received an email with this link, but can't find this Lars's post here and it looks like this repo is still private.
https://github.com/LarsArtmann/typespec-asyncapi

Any clarifications are welcome 😉

**Comment by LarsArtmann** [comment]
Hi,

I will release a Alpha v1 in hopefully the next 24h, but I am building it
with the help of Claude Code, so it was a bit fast with the Comment.
I made it public for you.

Best regards,
Lars Artmann
Managing Director at Artmann Holding GmbH and Artmann Technologies GmbH

Tel: +49 173 155 9729

**Comment by TertiumOrganum1** [comment]
That's awesome, thanks!

вс, 31 авг. 2025 г., 01:27 Lars Artmann ***@***.***>:

**Comment by LarsArtmann** [comment]
# 🎉 **TypeSpec AsyncAPI Emitter - Alpha Release Available**

I've implemented a working TypeSpec emitter for AsyncAPI 3.0 specifications, 
now available as `@lars-artmann/typespec-asyncapi@alpha`.

**Current Status:**
- ✅ Core functionality working: `@channel`, `@publish`, `@subscribe` decorators
- ✅ Complete TypeSpec model → AsyncAPI schema conversion
- ✅ Complex types supported: union types → enums, nested objects, optional fields
- ✅ AsyncAPI 3.0 specification compliance
- ✅ 100+ tests passing with comprehensive validation

**Installation & Usage:**
```bash
npm install @lars-artmann/typespec-asyncapi@alpha
```

Working example at: https://github.com/LarsArtmann/typespec-asyncapi/blob/master/test-example-advanced.tsp

Alpha Limitations: Advanced decorators like `@correlationId`, `@header` not yet implemented, but core event-driven API generation works. Find more infos on: https://www.npmjs.com/package/@lars-artmann/typespec-asyncapi or the dedicated GitHub repo.

💬 Please provide me als much and as harsh feedback on the current implementation as possible!

**Comment by CJSmith-0141** [comment]
> 💬 Please provide me als much and as harsh feedback on the current implementation as possible!

The signal to noise ratio of claude is clearly way off. This is completely unusable.

**Comment by LarsArtmann** [comment]
@CJSmith-0141 did you check the latest version on master or the https://github.com/LarsArtmann/typespec-asyncapi/tree/v0.0.1-alpha.2 version?

I agree, for now. What annoyed you the most?

**Comment by marvin-hsu** [comment]
Hi everyone, I recently used an AI tool to quickly build a working AsyncAPI emitter that we're already using internally at my company, hope you guys can give it a try!
- npm: https://www.npmjs.com/package/tsp-asyncapi
- docs: https://marvin-hsu.github.io/tsp-asyncapi/

**Comment by LarsArtmann** [comment]
It's been about a year since I last commented here with the alpha — @CJSmith-0141, you were right, it was unusable. I've since rewritten the whole thing.

`@lars-artmann/typespec-asyncapi@0.2.1-beta` is on npm now, targeting AsyncAPI 3.1. ~1300 tests, validating the output against the official 3.1 JSON Schema, binding definitions generated straight from `@asyncapi/specs`. Also `@typespec/versioning` support, multi-file output, and 22 protocols.

Nice to see @marvin-hsu's emitter here as well — good to have options.

@mario-guerra happy to share what I learned mapping AsyncAPI onto TypeSpec, or contribute the emitter itself if that's ever useful. 

@fmvilas your review offer still stands, I hope?

Repo: https://github.com/LarsArtmann/typespec-asyncapi
